Ai collab UI diffs (#24126)
microsoft/FluidFramework: packages/framework/ai-collab/README.md

### Using Trace Id's

Debug Events in ai-collab have two different types of trace id's:
- `traceId`: This field exists on all debug events and can be used to correlate all debug events that happened in a single execution of `aiCollab()`. Sorting the events by timestamp will show the proper chronological order of the events. Note that the events should already be emitted in chronological order.
- `eventFlowTraceId`: this field exists on all `EventFlowDebugEvents` and can be used to correlate all events from a particular event flow. Additionally all LLM api call events will contain the `eventFlowTraceId` field as well as a `triggeringEventFlowName` so you can link LLM API calls to a particular event flow.

> ## Edit Difference Visualizations

Reviewer: I don't think "visualizations" is appropriate either. I think this can just be "Edit Differences" or "Edit Diffs"

> ai-collab provides an array of `Diff` objects with its response. Each of these objects allows developers to identify tree nodes that have been modified as a result of ai collaboration and visualize them according to their needs.

Reviewer: Nit
```suggestion
`ai-collab` provides an array of `Diff` objects with its response. Each of these objects allows developers to identify tree nodes that have been modified as a result of ai collaboration and visualize them according to their needs.
```

Let's take a look at some examples for the following SharedTree application schema

```ts
import { aiCollab, DebugEvent, Diff } from "@fluidframework/ai-collab/alpha";

const sf = new SchemaFactory("testApp");

class Todo extends sf.object("Todo", {
	id: sf.identifier,
	title: string
	description: string
}) {}

class TestAppRootObject extends sf.object("TestAppRootObject", {
	id: sf.identifier,
	todos: sf.array([Todo]),
	innerObject: sf.object("InnerObject", {
		nestedTodos: sf.array([Todo]),
	}),
}) {}

const response = aiCollab({
	openAI: {
		client: new OpenAI({
			apiKey: OPENAI_API_KEY,
		}),
		modelName: "gpt-4o",
	},
	treeNode: view.root,
	prompt: {
		systemRoleContext:
			"You are a manager that is helping out with a project management tool. You have been asked to edit a group of tasks.",
		userAsk: userAsk,
	},
	limiters: {
		maxModelCalls: 25
	}
	planningStep: true,
	finalReviewStep: true,
	debugEventLogHandler: (event: DebugEvent) => {console.log(event);}
});

> const Diffs: diff[] = response.Diffs

Reviewer: The capitalization here seems backwards. Should this be:
```suggestion
const diffs: Diff[] = response.diffs;
```

Reviewer: This does not appear to have been fixed

```

> Each Diff will contain one or more `NodePath`'s. Each `NodePath` provides an array of objects that detail the path from the root node passed to ai-collab, down to the node targeted for editing. The first index in the `NodePath` is an object pointing to the target node and the last index is always the root node.

Reviewer: This paragraph is mostly repeating information given in the paragraph immediately before the example. We should deduplicate.

> Lets look at an example of the Insert Diff

Reviewer: ```suggestion
Let's look at an example of the Insert Diff.
```

> Lets look at an example of the Insert Diff

Reviewer: Nit: This and the following paragraph should probably go inside the `### Example Insert Ui Diff` subheading.

> The following `InsertDiff` is an example of a `Diff` that would result from if the ai agent inserts an object into index 1 of `TestAppRootObject.rootVectors`

Reviewer: ```suggestion
The following `InsertDiff` is an example of a `Diff` that would result from an AI agent inserting an object into index 1 of `TestAppRootObject.rootVectors`
```

### Example Insert Ui Diff

> ```json

Reviewer: Nit (blank line after heading)
```suggestion

```json
```

type: "insert",
nodePath: [
		{
			shortId: -14,
			schemaIdentifier: "testApp.Todo",
			parentField: 1,
		},
		{
			shortId: undefined,
			schemaIdentifier: "testApp.Array<[\"testApp.Todo\"]>",
			parentField: "todos",
		},
		{
			shortId: -1,
			schemaIdentifier: "testApp.TestAppRootObject",
			parentField: "rootFieldKey",
		},
	],
  	aiExplanation: "I need to insert a todo within the todos array",
  	nodeContent: {
		id: "f75951b0-df9d-4daa-aaf2-c322f2f462a8",
		title: "Example Todo Title",
		description: "Example Todo Description"
	},
```
As you can see, the object at the beginning of the `nodePath` array directly points to the newly inserted node while each next object is the parent of the preceding node until you hit the root node passed to the ai-collab function call.

Using the following ui diff, you can identify the modified node in a number of different ways.

> The simplest way is to use the `shortId` where you can use the following code to identify the newly inserted node within the SharedTree.

Reviewer: ```suggestion
The simplest way is to use the `shortId`, where you can use the following code to identify the newly inserted node within the SharedTree.
```

> **Note that the shortId field will only exist for objects that have a field defined as the `SchemaFactory.identifier` field.** See the above example app schema in this section to see the schema field defined as `sf.identifier`

Reviewer: You'll want to test this before merging, but GitHub supports a "notice" syntax for things like Notes and Warnings. I believe you should be able to do:
```suggestion
> [!Note]
> The `shortId` field will only exist for objects that have a field defined as the `SchemaFactory.identifier` field.

See the above example app schema in this section to see the schema field defined as `sf.identifier`
```

> Lets take a look at another UI example of using an array of Diffs to render changes

Reviewer: ```suggestion
Let's take a look at another UI example of using an array of Diffs to render changes.

```

```ts

> import { Tree } from "@fluidframework/tree/alpha"

Reviewer: Is this API alpha only? Or can we do:
```suggestion
import { Tree } from "@fluidframework/tree"
```

function renderTodoWithDiffs(todo: Todo, Diffs: diff[]) {
const modifyDiffs = Diffs.filter((diff): diff is ModifyDiff => diff.type === "modify") ?? [];
const matchingModifyDiffs = modifyDiffs.filter(
	(diff: ModifyDiff) =>
		// Modify diffs are a field level edit, so the first path will be the field on the target node and the second will be the node itself.
		diff.nodePath.length > 1 && diff.nodePath[1]?.shortId === Tree.shortId(task),
);

const insertDiffs = Diffs.filter((diff): diff is InsertDiff => diff.type === "insert") ?? [];
const matchingInsertDiffs = insertDiffs.filter(
	(diff: InsertDiff) =>
		// Insert diffs are a node level edit, so the first path will be the node.
		diff.nodePath[0]?.shortId === Tree.shortId(task),
);

> }

Reviewer: Looks like this function body isn't indented correctly, FYI

if (insertDiffs.length > 0) {
	renderNewlyInsertedTodo(todo)
} else if (modifyDiffs.length > 0) {
	const modifiedFields: [] = matchingModifyDiffs.map(diff => diff.nodePath[0].parentField);
	renderModifiedTodo(todo, modifiedFields)
} else {
	renderTodo(todo)
}

```

> You can also use the `type` and `schemaIdentifier` fields to group related `diff`'s

Reviewer: ```suggestion
You can also use the `type` and `schemaIdentifier` fields to group related `diff``s.
```

> ```
> You can also use the `type` and `schemaIdentifier` fields to group related `diff`'s
> ```ts

Reviewer: ```suggestion
```

You can also use the `type` and `schemaIdentifier` fields to group related `diff`'s.

```ts
```

> ```ts

Reviewer: Nit: there should be blank lines before and after fenced code blocks. GitHub won't let me make the suggestion because all of the backticks confuse it 🤣

- `eventFlowTraceId`: this field exists on all `EventFlowDebugEvents` and can be used to correlate all events from a particular event flow. Additionally all LLM api call events will contain the `eventFlowTraceId` field as well as a `triggeringEventFlowName` so you can link LLM API calls to a particular event flow.

## Edit Differences

ai-collab provides an array of `Diff` objects with its response. Each of these objects allows developers to identify tree nodes that have been modified as a result of AI collaboration and visualize them according to their needs.

Every `Diff` will include one or more `NodePaths` representing the nodes affected by a single edit created by the ai agent. A `NodePath` is an array whose items represent segment paths, beginning from the node targeted for modification (at the start of the array) all the way back to the root node passed to the ai-collab function call (at the end of the array), along with an explanation directly from the AI agent as to why it performed an edit.

> Let's take a look at some examples for the following SharedTree application schema

Reviewer: ```suggestion
Let's take a look at some examples for the following SharedTree application schema.
```

	},
	limiters: {
		maxModelCalls: 25
	}
	planningStep: true,
	finalReviewStep: true,
	debugEventLogHandler: (event: DebugEvent) => {console.log(event);}
});

const Diffs: diff[] = response.Diffs
```

Each Diff includes one or more NodePath objects. A `NodePath` is an array of objects representing the path from the root node (provided to ai-collab) to the specific node being edited. The array starts with the target node and ends with the root node.

Let's look at an example of the Insert Diff.

### Example Insert Diff

The following `InsertDiff` is an example of a `Diff` that would result from an AI agent inserting an object into index 1 of `TestAppRootObject.rootVectors`

```json
type: "insert",
nodePath: [
		{
			shortId: -14,
			schemaIdentifier: "testApp.Todo",
			parentField: 1,
		},
		{
			shortId: undefined,
			schemaIdentifier: "testApp.Array<[\"testApp.Todo\"]>",
			parentField: "todos",
		},
		{
			shortId: -1,
			schemaIdentifier: "testApp.TestAppRootObject",
			parentField: "rootFieldKey",
		},
	],
  	aiExplanation: "I need to insert a todo within the todos array",
  	nodeContent: {
Using a `Diff`, you can identify the modified node in a number of different ways.
```

> As you can see, the object at the beginning of the `nodePath` array directly points to the newly inserted node, while each subsequent object is the parent of the preceding node, terminating in the root node that was passed to the `ai-collab` function call.

Reviewer: Nit
```suggestion

As you can see, the object at the beginning of the `nodePath` array directly points to the newly inserted node, while each subsequent object is the parent of the preceding node, terminating in the root node that was passed to the `ai-collab` function call.
```

- `traceId`: This field exists on all debug events and can be used to correlate all debug events that happened in a single execution of `aiCollab()`. Sorting the events by timestamp will show the proper chronological order of the events. Note that the events should already be emitted in chronological order.
- `eventFlowTraceId`: this field exists on all `EventFlowDebugEvents` and can be used to correlate all events from a particular event flow. Additionally all LLM api call events will contain the `eventFlowTraceId` field as well as a `triggeringEventFlowName` so you can link LLM API calls to a particular event flow.

> ## Ui Visualizations

Reviewer: Nit: The diffs we return aren't specifically UI visualizations. Should we call this section "Edit Diffs" or something?

> ai-collab provides an array of `UiDiff` objects with its response.Each of these objects allows developers to identify tree nodes that have been modified as a result of ai collaboration and visualize them according to their needs.

Reviewer 2: ```suggestion
ai-collab provides an array of `UiDiff` objects with its response. Each of these objects allows developers to identify tree nodes that have been modified as a result of ai collaboration and visualize them according to their needs.
```

> ai-collab provides an array of `UiDiff` objects with its response. Each of these objects allows developers to identify tree nodes that have been modified as a result of ai collaboration and visualize them according to their needs.

Reviewer: Nit: markdown best practices call for a line break between heading and contents.
```suggestion

ai-collab provides an array of `UiDiff` objects with its response. Each of these objects allows developers to identify tree nodes that have been modified as a result of ai collaboration and visualize them according to their needs.
```

> ai-collab provides an array of `UiDiff` objects with its response. Each of these objects allows developers to identify tree nodes that have been modified as a result of ai collaboration and visualize them according to their needs.

Reviewer: Nit: our internal markdown best practices recommend putting line breaks between sentences in markdown. This helps scope future diffs, making docs changes a bit easier to review.

https://github.com/microsoft/FluidFramework/wiki/Markdown-Best-Practices#line-breaks-along-sentence-boundaries

> Every `UiDiff` will include either a single `NodePath` or multiple in the case of multiple nodes being targeted by a single edit created by the ai agent. A `NodePath` is an array whose items represent segment paths, beginning from the node targeted for modification (at the start of the array) all the way back to the root node passed to the ai-collab function call (at the end of the array), along with an explanation directly from the ai agent as to why it performed an edit.

Reviewer: This first sentence was a bit hard for me to parse. What do you think about something like
> Every `UiDiff` will include one or more `NodePath`s representing the nodes affected by a single edit created by the ai agent.

> Let's take a look at some examples for the following SharedTree application schema
> ```ts

Reviewer: Nit: markdown best practices call for line breaks on either side of a fenced code block. Also missing punctuation :)
```suggestion
Let's take a look at some examples for the following SharedTree application schema.

```ts
```

import { aiCollab, DebugEvent, UiDiff } from "@fluidframework/ai-collab/alpha";

const sf = new SchemaFactory("testApp");

class Todo extends sf.object("Todo", {
	id: sf.identifier,
	title: string
	description: string
}) {}

class TestAppRootObject extends sf.object("TestAppRootObject", {
	id: sf.identifier,
	todos: sf.array([Todo]),
	innerObject: sf.object("InnerObject", {
		nestedTodos: sf.array([Todo]),
	}),
}) {}

> const response = aiCollab({

Reviewer: I think this is still applicable?

	openAI: {
		client: new OpenAI({
			apiKey: OPENAI_API_KEY,
		}),
		modelName: "gpt-4o",
	},
	treeNode: view.root,
	prompt: {
		systemRoleContext:
			"You are a manager that is helping out with a project management tool. You have been asked to edit a group of tasks.",
		userAsk: userAsk,
	},
	limiters: {
		maxModelCalls: 25
	}
	planningStep: true,
	finalReviewStep: true,
	debugEventLogHandler: (event: DebugEvent) => {console.log(event);}
});

const uiDiffs: UiDiff[] = response.uiDiffs
```

Each UiDiff will contain one or more `NodePath`'s. Each `NodePath` provides an array of objects that detail the path from the root node passed to ai-collab, down to the node targeted for editing. The first index in the `NodePath` is an object pointing to the target node and the last index is always the root node.

> 

Reviewer: Nit (extra blank line between paragraphs)
```suggestion
```

Lets look at an example of the Insert Ui Diff
The following `InsertDiff` is an example of a `UiDiff` that would result from if the ai agent inserts an object into index 1 of `TestAppRootObject.rootVectors`
### Example Insert Ui Diff
```json
type: "insert",
nodePath: [
		{
			shortId: -14,
			schemaIdentifier: "testApp.Todo",
			parentField: 1,
		},
		{
			shortId: undefined,
			schemaIdentifier: "testApp.Array<[\"testApp.Todo\"]>",
			parentField: "todos",
		},
		{
			shortId: -1,
			schemaIdentifier: "testApp.TestAppRootObject",
			parentField: "rootFieldKey",
		},
	],
  	aiExplanation: "I need to insert a todo within the todos array",
  	nodeContent: {
		id: "f75951b0-df9d-4daa-aaf2-c322f2f462a8",
		title: "Example Todo Title",
		description: "Example Todo Description"
	},
```
As you can see, the object at the beginning of the `nodePath` array directly points to the newly inserted node while each next object is the parent of the preceding node until you hit the root node passed to the ai-collab function call.

> Using the following ui diff, you can identify the modified node in a number of different ways.

Reviewer: ```suggestion
Using a `Diff`, you can identify the modified node in a number of different ways.
```

		diff.nodePath[0]?.shortId === Tree.shortId(task),
);
}

if (insertDiffs.length > 0) {
	renderNewlyInsertedTodo(todo)
} else if (modifyDiffs.length > 0) {
	const modifiedFields: [] = matchingModifyDiffs.map(diff => diff.nodePath[0].parentField);
	renderModifiedTodo(todo, modifiedFields)
} else {
	renderTodo(todo)
}

```
You can also use the `type` and `schemaIdentifier` fields to group related `UiDiff`'s
```ts

const result = aiCollab({
	openAI: {
		client: new OpenAI({
			apiKey: OPENAI_API_KEY,
		}),
		modelName: "gpt-4o",
	},
	treeNode: view.root,
	prompt: {
		systemRoleContext:
			"You are a manager that is helping out with a project management tool. You have been asked to edit a group of tasks.",
		userAsk: userAsk,
	},
	limiters: {
		maxModelCalls: 25
	}
	planningStep: true,
	finalReviewStep: true,
	debugEventLogHandler: (event: DebugEvent) => {console.log(event);}
});

const uiDiffs: UiDiff[] = result.uiDiffs;

const insertDiffs = result.uiDiffs.filter(diff => diff.type === 'insert');

const insertedVectorsDiffs = insertDiffs.filter(diff => diff.path[0]?.schemaIdentifier === TestVector.identifier)

```

Other `UiDiff` types follow the same basic structure.

> Read the tsdoc [here](./src/aiCollabUiDiffApi.ts) for more info.

Reviewer: Nit: If possible, we should avoid linking to source code. It would be better to link to the API docs on fluidframework.com. That's a bit difficult in this case because the docs haven't been published yet. But the docs are published with a consistent theme, so we should be able to predict what the URL will be. For example, the `Diff` interface will end up at `https://fluidframework.com/docs/api/ai-collab/diff-interface`.

> Read the tsdoc [here](./src/aiCollabUiDiffApi.ts) for more info.

Reviewer: Given that `Diff` is a union of the collection of supported types, linking just to its docs is probably sufficient - its page will include references to all of its union members.

> [!NOTE]

> > The `shortId` field will only exist for objects that have a field defined as the `SchemaFactory.identifier` field.

Reviewer: Nit (line break after quote):
```suggestion
> The `shortId` field will only exist for objects that have a field defined as the `SchemaFactory.identifier` field.

```

> Let's take a look at another UI example of using an array of `Diff`s to render changes

Reviewer: ```suggestion
Let's take a look at another UI example of using an array of `Diff`s to render changes.
```
